include/envoy/buffer/buffer.h

```diff
@@ -142,9 +142,14 @@ class Factory {
 
   /**
    * Creates and returns a unique pointer to a new buffer.
+   * @param below_low_watermark supplies a function to call if the buffer goes under a configuerd
+   *   low watermark.
+   * @param above_high_watermark supplies a function to call if the buffer goes over a configured
+   *   high watermark.
    * @return a newly created InstancePtr.
    */
-  virtual InstancePtr create() PURE;
+  virtual InstancePtr create(std::function<void()> below_low_watermark,
```

Comment:
Sorry one other nit: Thoughts on renaming this class WatermarkFactory or renaming this function createWatermark(...) ? Even if it returns InstancePtr I think it would be more clear that it should be creating a watermark buffer vs. some other buffer that might ignore the watermark params.